File: julia/demosaic_data.jl
using Images
using PyPlot

img_array = Float64.(channelview(load("demosaic_raw.png")))
img_array = img_array[1:3,:,:];

_,m,n = size(img_array);
HEIGHT, WIDTH = m, n

# Set up the masks
R_mask = find([i%2==1 && j%2==1 for i in 1:m, j in 1:n])
G_mask = find([(i%2==1 && j%2==0) || (i%2==0 && j%2==1) for i in 1:m, j in 1:n])
B_mask = find([i%2==0 && j%2==0 for i in 1:m, j in 1:n])

# Load data correctly
R_raw = img_array[1,:,:];
G_raw = img_array[2,:,:];
B_raw = img_array[3,:,:];

function save_image(R,G,B)
  plt[:ioff]()
  I = cat(3,R,G,B)
  plt[:imshow](I)
  plt[:tick_params](axis="both",which="both", labelleft="off", labelbottom="off",
                                                  bottom="off", top="off", right="off", left="off")
  plt[:savefig]("output_image.png",bbox_inches="tight",pad_inches=0.)
end

File: python/demosaic_data.py
import numpy as np
import matplotlib.image as mpimg
import matplotlib.pyplot as plt

img_array = mpimg.imread('demosaic_raw.png')[:,:,:3]
HEIGHT, WIDTH, _ = img_array.shape

m, n = HEIGHT, WIDTH

# Set up the masks
R_mask = np.where([[i%2==0 and j%2==0 for j in range(WIDTH)] for i in range(HEIGHT)])
G_mask = np.where([[(i%2==0 and j%2==1) or (i%2==1 and j%2==0) for j in range(WIDTH)] for i in range(HEIGHT)])
B_mask = np.where([[i%2==1 and j%2==1 for j in range(WIDTH)] for i in range(HEIGHT)])

# Load the data correctly
R_raw = np.zeros((m, n))
G_raw = np.zeros((m, n))
B_raw = np.zeros((m, n))

R_known = img_array[:,:,0]
G_known = img_array[:,:,1]
B_known = img_array[:,:,2]

R_raw[R_mask] = R_known[R_mask]
G_raw[G_mask] = G_known[G_mask]
B_raw[B_mask] = B_known[B_mask]

def save_image(R,G,B,name='output_image.png'):
    img = np.stack((np.array(R),np.array(G),np.array(B)), axis=2)
    # turn off ticks and labels of the figure
    plt.tick_params(
      axis='both', which='both', labelleft='off', labelbottom='off',
      bottom='off', top='off', right='off', left='off'
    )
    fig = plt.imshow(img)
    plt.savefig(name,bbox_inches='tight',pad_inches=0.)
